src/main/java/com/google/googleidentity/oauth2/endpoint/TokenEndpoint.java

```diff
@@ -0,0 +1,139 @@
+                            "Error Type: " + exception.getErrorType() +
+                            "Description: " + exception.getErrorDescription());
+            response.setStatus(exception.getHttpCode());
+            response.setContentType("application/json;charset=UTF-8");
+            response.getWriter().println(
+                    OAuth2ExceptionHandler.getResponseBody(exception).toJSONString());
+            response.getWriter().flush();
+            return;
+        }
+
+        OAuth2Request oauth2Request = parseOAuth2RequestFromHttpRequest(request);
+    }
+
+    /**
+     * Should be called after all validation in doPost function.
+     */
+    private OAuth2Request parseOAuth2RequestFromHttpRequest(HttpServletRequest request) {
+
+        String grantType = request.getParameter(OAuth2ParameterNames.GRANT_TYPE);
+
+        Preconditions.checkArgument(
+                OAuth2Utils.getClientSession(request).getClient().isPresent(),
+                "Client should have been set in client filter!");
+
+        OAuth2Request.Builder oauth2RequestBuilder = OAuth2Request.newBuilder();
+        oauth2RequestBuilder.getRequestAuthBuilder().setClientId(
+                            OAuth2Utils.getClientSession(request).getClient().get().getClientId());
+        oauth2RequestBuilder.getRequestBodyBuilder()
+                .setGrantType(grantType).setResponseType("token");
+
+        if (grantType.equals("authorization_code")) {
+            oauth2RequestBuilder.getRequestAuthBuilder().setCode(
+                    request.getParameter(OAuth2ParameterNames.CODE));
+            try {
+                oauth2RequestBuilder.getAuthorizationResponseBuilder().setRedirectUri(
+                        URLDecoder.decode(
+                                request.getParameter(OAuth2ParameterNames.REDIRECT_URI),
+                                "utf-8"));
+            } catch (UnsupportedEncodingException e) {
+                // This should never happen.
+                throw new IllegalStateException(
+                        "URL should be valid", e);
+            }
+
+        } else if (grantType.equals("refresh_token")) {
+            oauth2RequestBuilder.getRequestBodyBuilder().setRefreshToken(
+                    request.getParameter(OAuth2ParameterNames.REFRESH_TOKEN));
+        } else if (grantType.equals("urn:ietf:params:oauth:grant-type:jwt-bearer")) {
+            oauth2RequestBuilder.getRequestBodyBuilder()
+                    .setIntent(request.getParameter(OAuth2ParameterNames.INTENT))
+                    .setAssertion(request.getParameter(OAuth2ParameterNames.ASSERTION));
+            if (!Strings.isNullOrEmpty(request.getParameter(OAuth2ParameterNames.SCOPE))) {
+                oauth2RequestBuilder.getAuthorizationResponseBuilder().setState(
+                        request.getParameter(OAuth2ParameterNames.SCOPE));
+            }
+            log.info("Assertion: " + request.getParameter(OAuth2ParameterNames.ASSERTION));
```

Comment:
Assertions have PII in them (the user's name and email among other things). I would recommend against logging them explicitly like this. If anything, maybe only do it at some finer log level than "info" so that it's not on by default.